pandas/tests/frame/test_alter_axes.py

```diff
@@ -392,6 +392,11 @@ def test_dti_set_index_reindex(self):
 
         assert new_index.freq == index.freq
 
+    def test_set_index_tuple(self):
+        # test for fix of issue #15457, where the following raised a TypeError
+        df = DataFrame([[2, 1, 2], [4, (1, 2), 3]]).set_index([0, 1])
+        assert df.to_dict() == {2: {(2, 1): 2, (4, (1, 2)): 3}}
```

Comment:
use assert_frame_equal, IOW you must construct the expected frame. use ``result=`` and ``expected=`` here

Reply:
Comparing from_arrays with from_tuples? Not sure I understand. As you said, this already works.

Reply:
@toobaz I reverted to the test via dict as otherwise line 908 of cast.py causes an error when trying to create an array from [1, (1, 2)], raising {ValueError}setting an array element with a sequence. This may be an unrelated issue but not sure the test can be constructed in that way at the moment.

Reply:
@dickreuter this has to be an assert on a frame, otherwise this test is not testing what we need here

Reply:
> {ValueError}setting an array element with a sequence

@dickreuter You are right, sorry. This should work instead:
```
expected = pd.DataFrame([[2, 1], [4, (1, 2)]]).set_index([0, 1]).index
result = pd.MultiIndex.from_tuples([(2, 1), (4, (1,2))], names=(0, 1))
```

@jreback see my comment below, the bug is unrelated to ``DataFrame``